src/main/scala/kamon/datadog/DatadogAgentReporter.scala

```diff
@@ -100,25 +100,111 @@ class DatadogAgentReporter extends MetricReporter {
     case _                                                            => value.toDouble
   }
 
-  private def readConfiguration(config: Config): Configuration = {
+}
+
+object DatadogAgentReporter {
+
+  private val logger = LoggerFactory.getLogger(classOf[DatadogAgentReporter])
+
+  trait MeasurementFormatter {
+    def formatMeasurement(measurementData: String, tags: Map[String, String]): String
+  }
+
+  private class DefaultMeasurementFormatter(config: Config) extends MeasurementFormatter {
+
+    private val tagConfig = readTagConfig(config)
+
+    override def formatMeasurement(
+      measurementData: String,
+      tags: Map[String, String]
+    ): String  = {
+
+      val filteredTags = Map(tagConfig.serviceTagName -> Kamon.environment.service) ++ tags.filterKeys {
+        case k =>
+          tagConfig.filters.includes.exists(_.pattern.matcher(k).matches()) &&
+          !tagConfig.filters.excludes.exists(_.pattern.matcher(k).matches())
+      }
+
+      val stringTags: String = "#|" + filteredTags.map { case (k, v) ⇒ k + ":" + v }.mkString(",")
+
+      StringBuilder.newBuilder
+      .append(measurementData)
+      .append(stringTags)
+      .result()
+    }
+  }
+
+
+  private[datadog] def readConfiguration(config: Config): Configuration = {
     val datadogConfig = config.getConfig("kamon.datadog")
 
     Configuration(
-      agentAddress = new InetSocketAddress(datadogConfig.getString("agent.hostname"), datadogConfig.getInt("agent.port")),
-      maxPacketSize = datadogConfig.getBytes("agent.max-packet-size"),
       timeUnit = readTimeUnit(datadogConfig.getString("time-unit")),
-      informationUnit = readInformationUnit(datadogConfig.getString("information-unit")))
+      informationUnit = readInformationUnit(datadogConfig.getString("information-unit")),
+      measurementFormatter = getMeasurementFormatter(datadogConfig),
+      packetBuffer = getPacketBuffer(datadogConfig)
+    )
   }
 
-  private case class Configuration(agentAddress: InetSocketAddress, maxPacketSize: Long, timeUnit: MeasurementUnit,
-                                   informationUnit: MeasurementUnit)
 
-  private class PacketBuffer(maxPacketSizeInBytes: Long, channel: DatagramChannel, remote: InetSocketAddress) {
+  // Copied from akka
+  private def getClassFor[T: ClassTag](fqcn: String): Try[Class[_ <: T]] =
```

Comment:
This whole section is already in Kamon (also from Akka :smiling_imp:), please take a look at this: https://github.com/kamon-io/Kamon/blob/master/kamon-core/src/main/scala/kamon/util/DynamicAccess.scala